src/FSharp.Data.GraphQL.Server.AppInfrastructure/GraphQLWebsocketMiddleware.fs

```diff
@@ -0,0 +1,375 @@
+
+  let handleMessages (cancellationToken: CancellationToken) (serializerOptions: JsonSerializerOptions) (executor : Executor<'Root>) (root: unit -> 'Root) (pingHandler : PingHandler option) (socket : WebSocket) =
+    let subscriptions = new Dictionary<SubscriptionId, SubscriptionUnsubscriber * OnUnsubscribeAction>()
+    // ---------->
+    // Helpers -->
+    // ---------->
+    let safe_ReceiveMessageViaSocket = receiveMessageViaSocket (new CancellationToken())
+
+    let safe_Send = sendMessageViaSocket serializerOptions socket
+    let safe_Receive() =
+      socket
+      |> safe_ReceiveMessageViaSocket serializerOptions executor
+
+    let safe_SendQueryOutput id output =
+      let outputAsDict = output :> IDictionary<string, obj>
+      match outputAsDict.TryGetValue("errors") with
+      | true, theValue ->
+        // The specification says: "This message terminates the operation and no further messages will be sent."
+        subscriptions
+        |> GraphQLSubscriptionsManagement.removeSubscription(id)
+        safe_Send (Error (id, unbox theValue))
+      | false, _ ->
+        safe_Send (Next (id, output))
+
+    let sendQueryOutputDelayedBy (cancToken: CancellationToken) (ms: int) id output =
+      task {
+            do! Async.StartAsTask(Async.Sleep ms, cancellationToken = cancToken)
+            do! output
+                |> safe_SendQueryOutput id
+        }
+    let safe_SendQueryOutputDelayedBy = sendQueryOutputDelayedBy cancellationToken
+
+    let safe_ApplyPlanExecutionResult (id: SubscriptionId) (socket) (executionResult: GQLResponse)  =
+      task {
+            match executionResult with
+            | Stream observableOutput ->
+                subscriptions
+                |> addClientSubscription id serializerOptions socket safe_SendQueryOutput observableOutput
+            | Deferred (data, errors, observableOutput) ->
+                do! data
+                    |> safe_SendQueryOutput id
+                if errors.IsEmpty then
+                  subscriptions
+                  |> addClientSubscription id serializerOptions socket (safe_SendQueryOutputDelayedBy 5000) observableOutput
+                else
+                  ()
+            | Direct (data, _) ->
+                do! data
+                    |> safe_SendQueryOutput id
+        }
+
+    let getStrAddendumOfOptionalPayload optionalPayload =
+        optionalPayload
+        |> Option.map (fun payloadStr -> sprintf " with payload: %A" payloadStr)
+        |> Option.defaultWith (fun () -> "")
+
+    let logMsgReceivedWithOptionalPayload optionalPayload (msgAsStr : string) =
+      if logger.IsEnabled(LogLevel.Trace) then
+        logger.LogTrace ("{message}{messageaddendum}", msgAsStr, (optionalPayload |> getStrAddendumOfOptionalPayload))
+
+    let logMsgWithIdReceived (id : string) (msgAsStr : string) =
+      if logger.IsEnabled(LogLevel.Trace) then
+        logger.LogTrace("{message} (id: {messageid})", msgAsStr, id)
```

Comment:
```suggestion
      logger.LogTrace("{message} (id: {messageid})", msgAsStr, id)
```

Reply:
Done by yourself, so I'm resolving the conversation.